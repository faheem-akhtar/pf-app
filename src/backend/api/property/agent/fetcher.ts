import { BackendApiFactory } from 'backend/api/factory';
import { BackendApiPropertyJsonApiResultType } from '../json-api-result.type';
import { PropertyAgentResultType } from 'components/property/agent-result.type';

import { backendApiPropertyAgentMapper } from './mapper';

const fetcher = BackendApiFactory<PropertyAgentResultType, BackendApiPropertyJsonApiResultType>({
  method: 'GET',
  url: 'property',
  queryDefaultParams: {
    include: 'agent,agent.languages',
  },
  dataMapper: backendApiPropertyAgentMapper,
});

export const backendApiPropertyAgentFetcher = (locale: string, id: string): ReturnType<typeof fetcher> => {
  return fetcher({
    locale,
    query: {
      'filter[ids]': [id],
    },
  });
};
