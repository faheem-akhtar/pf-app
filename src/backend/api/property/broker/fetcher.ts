import { BackendApiFactory } from 'backend/api/factory';
import { PropertyBrokerInterface } from 'components/property/broker.interface';
import { configOriginIfDevUseStagingValue } from 'config/origin/if-dev-use-staging-value';

import { BackendApiPropertyJsonApiResultType } from '../json-api-result.type';
import { backendApiPropertyBrokerMapper } from './mapper';

const fetcher = BackendApiFactory<PropertyBrokerInterface, BackendApiPropertyJsonApiResultType>({
  method: 'GET',
  url: 'property',
  queryDefaultParams: {
    include: 'broker',
  },
  dataMapper: backendApiPropertyBrokerMapper,
});

export const backendApiPropertyBrokerFetcher = (locale: string, id: string): ReturnType<typeof fetcher> => {
  return fetcher({
    locale,
    getOrigin: configOriginIfDevUseStagingValue,
    query: {
      'filter[ids]': [id],
    },
  });
};
