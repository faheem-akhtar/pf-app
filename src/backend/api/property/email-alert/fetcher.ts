import { BackendApiFactory } from 'backend/api/factory';
import { BackendJsonApiModelType } from 'backend/json-api/model.type';
import { EmailAlertAttributesInterface } from 'types/email-alert/attributes.interface';

import { configOriginIfDevUseStagingValue } from 'config/origin/if-dev-use-staging-value';

const fetcher = BackendApiFactory<BackendJsonApiModelType, BackendJsonApiModelType>({
  method: 'POST',
  url: 'email-alert',
  alterHeaders: (headers) => {
    headers['content-type'] = 'application/vnd.api+json';
  },
});

export const backendApiPropertyEmailAlertFetcher = (
  locale: string,
  attributes: EmailAlertAttributesInterface
): ReturnType<typeof fetcher> =>
  fetcher({
    locale,
    getOrigin: configOriginIfDevUseStagingValue,
    postData: {
      data: {
        type: 'email_alert',
        attributes,
      },
    },
  });
