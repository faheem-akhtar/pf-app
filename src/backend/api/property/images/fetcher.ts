import { BackendApiFactory } from 'backend/api/factory';

import { BackendApiPropertyJsonApiResultType } from '../json-api-result.type';
import { backendApiPropertyImagesMapper } from './mapper';
import { BackendApiPropertyImagesType } from './type';

const fetcher = BackendApiFactory<BackendApiPropertyImagesType, BackendApiPropertyJsonApiResultType>({
  method: 'GET',
  url: 'property',
  queryDefaultParams: {
    'filter[allow_expired]': true,
    include: 'property_images',
  },
  dataMapper: backendApiPropertyImagesMapper,
});

export const backendApiPropertyImagesFetcher = (locale: string, ids: string[]): ReturnType<typeof fetcher> => {
  return fetcher({
    locale,
    query: {
      'filter[allow_expired]': true,
      'filter[ids]': ids,
      'page[limit]': ids.length,
      include: 'property_images',
    },
  });
};
