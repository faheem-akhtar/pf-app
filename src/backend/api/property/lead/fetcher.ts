import { BackendApiFactory } from 'backend/api/factory';
import { BackendJsonApiModelType } from 'backend/json-api/model.type';
import { configOriginIfDevUseStagingValue } from 'config/origin/if-dev-use-staging-value';
import { PropertyLeadAttributesInterface } from 'types/property/lead/attributes.interface';

type getFetcherReturnType = ReturnType<typeof BackendApiFactory>;

const getFetcher = (url: string): getFetcherReturnType =>
  BackendApiFactory<BackendJsonApiModelType, BackendJsonApiModelType>({
    method: 'POST',
    url,
    alterHeaders: (headers) => {
      headers['content-type'] = 'application/vnd.api+json';
    },
  });

export const backendApiPropertyLeadFetcher = (
  locale: string,
  propertyId: string,
  attributes: PropertyLeadAttributesInterface
): ReturnType<getFetcherReturnType> => {
  const fetcher = getFetcher(`property/${propertyId}/lead`);

  return fetcher({
    locale,
    getOrigin: configOriginIfDevUseStagingValue,
    postData: {
      data: {
        type: 'property_lead',
        attributes,
      },
    },
  });
};
