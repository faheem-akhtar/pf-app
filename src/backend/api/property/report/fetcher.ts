import { ApiPropertyReportRequestInterface } from 'api/property/report/request.interface';
import { BackendApiFactory } from 'backend/api/factory';
import { BackendJsonApiModelType } from 'backend/json-api/model.type';
import { configOriginIfDevUseStagingValue } from 'config/origin/if-dev-use-staging-value';

const fetcher = BackendApiFactory<BackendJsonApiModelType, BackendJsonApiModelType>({
  method: 'POST',
  url: '',
  alterHeaders: (headers) => {
    headers['content-type'] = 'application/vnd.api+json';
    delete headers['Host'];
  },
});

export const backendApiPropertyReportFetcher = (
  locale: string,
  propertyId: string,
  attributes: ApiPropertyReportRequestInterface
): ReturnType<ReturnType<typeof BackendApiFactory>> => {
  return fetcher({
    locale,
    getOrigin: configOriginIfDevUseStagingValue,
    url: `property/${propertyId}/report`,
    postData: {
      data: {
        type: 'property_report',
        attributes,
      },
    },
  });
};
