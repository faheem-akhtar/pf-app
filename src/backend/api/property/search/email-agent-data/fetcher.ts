import { BackendApiFactory } from 'backend/api/factory';

import { BackendApiPropertySearchJsonApiResultType } from '../json-api-result.type';
import { backendApiPropertySearchEmailAgentDataMapper } from './mapper';
import { BackendApiPropertySearchEmailAgentDataResultType } from './result-type';

const fetcher = BackendApiFactory<
  BackendApiPropertySearchEmailAgentDataResultType,
  BackendApiPropertySearchJsonApiResultType
>({
  method: 'GET',
  url: 'search',
  queryDefaultParams: {
    include: 'properties,properties.location_tree,properties.property_type',
  },
  dataMapper: backendApiPropertySearchEmailAgentDataMapper,
});

export const backendApiPropertySearchEmailAgentDataFetcher = (
  locale: string,
  propertyId: string
): ReturnType<typeof fetcher> => {
  return fetcher({
    locale,
    query: {
      'filter[ids]': [propertyId],
      'page[limit]': 1,
    },
  });
};
