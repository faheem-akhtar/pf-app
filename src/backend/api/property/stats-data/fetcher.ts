import { BackendApiFactory } from 'backend/api/factory';
import { PropertyObfuscatedStatsDataResultType } from 'components/property/obfuscated-stats-data-result.type';
import { configIsTrace } from 'config/is-trace';

import { BackendApiPropertySearchJsonApiResultType } from '../search/json-api-result.type';
import { BackendApiPropertyStatsDataFetcherPropsType } from './fetcher-props.type';
import { backendApiPropertyStatsDataMapper } from './mapper';

const fetcher = BackendApiFactory<
  { properties: PropertyObfuscatedStatsDataResultType[] },
  BackendApiPropertySearchJsonApiResultType
>({
  method: 'GET',
  url: 'search',
});

const fieldsToInclude = [
  'properties',
  'properties.property_type',
  'properties.property_images',
  'properties.location_tree',
  'properties.agent',
  'properties.agent.languages',
  'properties.broker',
  'properties.project',
  'properties.project.developer',
  'properties.project.images',
  'properties.project_property',
  'properties.project_property.unit',
  'properties.project_property.unit.floor_plans',
  'smart_ads',
  'smart_ads.agent',
  'smart_ads.broker',
  'smart_ads.property_type',
  'smart_ads.property_images',
  'smart_ads.location_tree',
  'direct_from_developer',
  'direct_from_developer.property_type',
  'direct_from_developer.property_images',
  'direct_from_developer.location_tree',
  'direct_from_developer.agent',
  'direct_from_developer.broker',
  'cts',
  'cts.agent',
  'cts.broker',
  'cts.property_type',
  'cts.property_images',
  'cts.location_tree',
  'similar_properties',
  'similar_properties.agent',
  'similar_properties.broker',
  'similar_properties.property_type',
  'similar_properties.property_images',
  'similar_properties.location_tree',
  'agent_smart_ads',
  'agent_smart_ads.broker',
  'agent_smart_ads.languages',
  'agent_properties_smart_ads',
  'agent_properties_smart_ads.agent',
  'agent_properties_smart_ads.broker',
  'agent_properties_smart_ads.location_tree',
  'agent_properties_smart_ads.property_type',
  'agent_properties_smart_ads.property_images',
];

export const backendApiPropertyStatsDataFetcher = (
  props: BackendApiPropertyStatsDataFetcherPropsType
): ReturnType<typeof fetcher> => {
  const { locale, propertiesIds } = props;

  return fetcher({
    locale,
    query: {
      'filter[ids]': propertiesIds,
      'page[limit]': propertiesIds.length,
      include: fieldsToInclude.join(','),
    },
    dataMapper: (properties, rawResponse) => {
      const obfuscatedProperties = backendApiPropertyStatsDataMapper(
        properties as BackendApiPropertySearchJsonApiResultType,
        props
      );

      const result = {
        properties: obfuscatedProperties,
      };

      if (configIsTrace) {
        (result as Record<string, object>).__full = JSON.parse(JSON.stringify(rawResponse));
      }

      return result;
    },
  });
};
