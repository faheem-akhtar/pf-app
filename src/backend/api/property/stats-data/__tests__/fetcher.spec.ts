import { mockWindowFetch } from 'mocks/window/fetch.mock';
import { propertiesRawBackendStub } from 'stubs/properties/raw-backend.stub';

import { backendApiPropertyStatsDataFetcher } from '../fetcher';

describe('backendApiPropertyStatsDataFetcher', () => {
  it('should create correct url query', () => {
    const fetchMock = mockWindowFetch({ json: () => Promise.resolve(propertiesRawBackendStub()) });
    backendApiPropertyStatsDataFetcher({
      locale: 'en',
      propertiesIds: ['1', '2'],
      pageNumber: 30,
    });

    expect(fetchMock).toHaveBeenCalledTimes(1);
    expect(fetchMock).toHaveBeenCalledWith(
      'http://website-pf-local/en/api/search?filter%5Bids%5D%5B%5D=1&filter%5Bids%5D%5B%5D=2&page%5Blimit%5D=2&include=properties%2Cproperties.property_type%2Cproperties.property_images%2Cproperties.location_tree%2Cproperties.agent%2Cproperties.agent.languages%2Cproperties.broker%2Cproperties.project%2Cproperties.project.developer%2Cproperties.project.images%2Cproperties.project_property%2Cproperties.project_property.unit%2Cproperties.project_property.unit.floor_plans%2Csmart_ads%2Csmart_ads.agent%2Csmart_ads.broker%2Csmart_ads.property_type%2Csmart_ads.property_images%2Csmart_ads.location_tree%2Cdirect_from_developer%2Cdirect_from_developer.property_type%2Cdirect_from_developer.property_images%2Cdirect_from_developer.location_tree%2Cdirect_from_developer.agent%2Cdirect_from_developer.broker%2Ccts%2Ccts.agent%2Ccts.broker%2Ccts.property_type%2Ccts.property_images%2Ccts.location_tree%2Csimilar_properties%2Csimilar_properties.agent%2Csimilar_properties.broker%2Csimilar_properties.property_type%2Csimilar_properties.property_images%2Csimilar_properties.location_tree%2Cagent_smart_ads%2Cagent_smart_ads.broker%2Cagent_smart_ads.languages%2Cagent_properties_smart_ads%2Cagent_properties_smart_ads.agent%2Cagent_properties_smart_ads.broker%2Cagent_properties_smart_ads.location_tree%2Cagent_properties_smart_ads.property_type%2Cagent_properties_smart_ads.property_images',
      expect.anything()
    );
  });

  it('should map data correctly and obfuscate the keys', async () => {
    mockWindowFetch({ json: () => Promise.resolve(propertiesRawBackendStub()) });
    const result = await backendApiPropertyStatsDataFetcher({
      locale: 'en',
      propertiesIds: ['1', '2'],
      pageNumber: 30,
    });

    expect(result.ok).toBeTruthy();
    if (result.ok) {
      expect(result.data.properties).toMatchSnapshot();
    }
  });
});
