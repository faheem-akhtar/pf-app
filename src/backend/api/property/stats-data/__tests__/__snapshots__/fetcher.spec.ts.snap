// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backendApiPropertyStatsDataFetcher should map data correctly and obfuscate the keys 1`] = `
Array [
  Object {
    "006501075e65": 61.19,
    "01651115590d1c0c1d140374": "1",
    "04651a07681d0d0f070a19020c6f": false,
    "04651f1d4e000010172b1506086f": "",
    "056501075d65": 89.69,
    "076510015f170b0d110070": "AED",
    "0d651f1d5e11070d152a04151d1a1d6902160918040a114965": "standard",
    "0e651213480b1a63": Object {
      "32651a077b001c0a1410151069": false,
      "41651d1540006e": "Mr Hamid",
      "45651a102d52": 156740,
    },
    "0e651b155e5658532410150369": false,
    "106501114b001c061c1a153a1c020c481965": "FURNISH 33K",
    "10651213480b1a36011c023d0d6f": 28140,
    "1165031b5e0c1a0a1d1720150e0a6e": 1,
    "1165101b4015020606101f1a3a1b0f591e1679": null,
    "11651a194c020b63": "https://www.propertyfinder.ae/property/3656619107f9c0c1f5eb8ae1539b9c81/500/356/MODE/2627b3/7750053-d109eo.jpg?ctr=ae",
    "1765030644060b63": 3300,
    "17651f1d5e11070d152a04151d1a1d2d": "standard",
    "186505115f0c080a171d70": "Not verified",
    "186511114917010c1f0a70": "0",
    "1a6502072d11": 100,
    "1a651a077e080f1106381474": 0,
    "1f65030644060b370a0d70": "3,300 AED/month",
    "2365071d59090b63": "Family Community Furnish Studio/3300 M W/Markert",
    "24651b1b5816070d152d09040c260a2d": 1,
    "2765001d57006e": 650,
    "2765031b5e0c1a0a1d1770": 726,
    "2c65141159351c0a111c3f1a281f1e2d": false,
    "2e6506064165": "https://www.propertyfinder.ae/en/rent/apartment-for-rent-abu-dhabi-khalifa-city-khalifa-city-a-7750053.html",
    "3565011143110f0f221c021d060b6e": "Monthly",
    "35651d115a351c0c181c13001a6f": false,
    "37651b1b5816070d152d09040c6f": "apartment",
    "37651e1b5f110902151c2211071b6e": false,
    "38651106420e0b1172": Object {
      "26651d1540006e": "Moon Star Real Estate",
      "3a651a102d31": 2867,
    },
    "3b65011143110f0f221c021d060b27496b": "monthly",
    "3b651a076c170b0221091517000e0244181129060a03115f111763": 0,
    "3b651a102d30": 7750053,
    "3c651f1b4e041a0a1d170374": Object {
      "3d654374632a": 6,
      "3e6542746235": 299,
      "3f6541747d34": 1933,
    },
    "3f651e1b5b003d0c070870": false,
    "446500014f1107171e1c70": "Apartment for Rent in Khalifa City A, Khalifa City",
    "44651a076e111d33001600111b1b172d": 0,
    "45651c124b001c0a1c1e240d190a6e": "rent",
    "466515015f0b07101a1c1474": "YES",
    "4c651a194c020b1072": 10,
  },
]
`;
